### DISCUSSION TITLE
Best way to update other state when an input's state updates.

### DISCUSSION THREAD

**Discussion by SCasarotto** [ariakit/ariakit#1249] [Questions] [answered]
I am converting an app to use Ariakit and I ran into a use case that I wasn't sure how to best handle. My goal is to use `useSelectState` but also be able to update other state values. This could be done with an `useEffect` but that will cause double rendering for each state update. I noticed that you can pass `setValue` into the `useSelectState` options but the issue is that the other state is reliant on more than just this select state. After some trial and error, I found that I could do the following but it feels really clunky. 

```tsx
const modeSelectState = useSelectState<typeof modeOptions[number]>({
  defaultValue: '6 Hours',
  gutter: 4,
  sameWidth: true,
})
const [startOfDayTime, setStartOfDayTime] = useState<number | undefined>(
  addHours(startOfDay(new Date()), 8).getTime(),
)

<div>
  <SelectLabel state={modeSelectState}>Treatment Ratio</SelectLabel>
  <Select
    state={modeSelectState}
  />
  <SelectPopover
  // This feels ugly and weird but did deliver the results I was looking for.
  state={{
    ...modeSelectState,
    setValue: (value: string) => {
      modeSelectState.setValue(value)
      setHourArray((prev) =>
        //  A helper function. Not important for this example
        recalculateHours({
          hourArray: prev,
          scale: modeScales[value as typeof modeOptions[number]],
          startOfDayTime,
        }),
      )
    },
  }}
  >
  {modeOptions.map((option) => (
    <SelectItem
      key={option}
      value={option}
    >
      {option}
    </SelectItem>
  ))}
  </SelectPopover>
</div>

```

I am curious what the best solution here might be.

**Comment by diegohaz** [comment, marked as answer]
The example is only using `value` from the select state, but you mentioned you need to access more props. Have you tried passing `setValue` as an argument to `useSelectState` and accessing those other props there?

By re-creating the state object, you'll lose some performance optimizations. But it's not a big deal unless you have an extensive list (100+ items). This might be relevant, though, since you're trying to avoid re-renders.

Anyway, if you're going to do that, you probably want to create that state object outside of `SelectPopover` and pass it to all select components that receive the state prop to avoid any inconsistency (for example, `Select` may call `setValue`, which your change would miss).

However, I'd instead use the `useEffect` approach. That's the safest and most idiomatic approach.

But if you really want to avoid re-renders (although you probably shouldn't worry too much about this), you can also follow this example: https://reactjs.org/docs/hooks-faq.html#how-do-i-implement-getderivedstatefromprops

We're using this technique in a few places:

https://github.com/ariakit/ariakit/blob/031b639c7315a74826203cc121098d69208b5511/packages/ariakit/src/disclosure/disclosure-state.ts#L38-L40

https://github.com/ariakit/ariakit/blob/031b639c7315a74826203cc121098d69208b5511/packages/ariakit/src/combobox/combobox-state.ts#L105-L107

We really need those states to be in sync to avoid invalid UI states in both cases. We're not trying to prevent re-renders.

**Reply by SCasarotto** [reply]
Okay, thanks for the reply @diegohaz . I had a feeling the `useEffect` method was going to come out as the best option. For 1, it's much easier to think about it and keeps things simple. 

Just to make sure I understand. The option _without_ the use of `useEffect` would look something like:

```
const modeSelectState = useSelectState<typeof modeOptions[number]>({
  defaultValue: '6 Hours',
  gutter: 4,
  sameWidth: true,
})
const [startOfDayTime, setStartOfDayTime] = useState<number | undefined>(
  addHours(startOfDay(new Date()), 8).getTime(),
)
const overriddenModeSelectState = useMemo(() => ({
    ...modeSelectState,
    setValue: (value: string) => {
      modeSelectState.setValue(value)
      setHourArray((prev) =>
        //  A helper function. Not important for this example
        recalculateHours({
          hourArray: prev,
          scale: modeScales[value as typeof modeOptions[number]],
          startOfDayTime,
        }),
      )
    },
  }), 
  [modeSelectState, startOfDayTime]
)

<div>
  <SelectLabel state={overriddenModeSelectState}>Treatment Ratio</SelectLabel>
  <Select
    state={overriddenModeSelectState}
  />
  <SelectPopover
  // This feels ugly and weird but did deliver the results I was looking for.
  state={overriddenModeSelectState}
  >
  {modeOptions.map((option) => (
    <SelectItem
      key={option}
      value={option}
    >
      {option}
    </SelectItem>
  ))}
  </SelectPopover>
</div>
```

Is that correct?

**Reply by diegohaz** [reply]
Yes! But, as I said, I wouldn't recommend that.
